File: 19th/코딩테스트 심화반/week6/18126번_너구리 구구.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<pair<int, int>> adj[5001]; // 그래프를 인접리스트로 저장
bool visited[5001]; // 방문 여부

long long max_dist = 0; // 가장 먼 방까지의 거리

void dfs(int n, long long dist) {
    for (int i = 0; i < adj[n].size(); i++) {
        if (!visited[adj[n][i].first]) { // 방문하지 않았을 경우
            if (max_dist < dist + 1LL * adj[n][i].second) { // 현재까지의 거리가 더 클 경우
                max_dist = dist + 1LL * adj[n][i].second;
            }

            visited[adj[n][i].first] = true;
            dfs(adj[n][i].first, dist + 1LL * adj[n][i].second);
        }
    }
}

int main() {
    int N;
    cin >> N;

    int A, B, C;
    for (int i = 0; i < N - 1; i++) {
        cin >> A >> B >> C;
        adj[A].push_back({B, C});
        adj[B].push_back({A, C});
    }

    visited[1] = true;
    dfs(1, 0);

    cout << max_dist;
}

File: 19th/코딩테스트 심화반/week6/18126번_너구리 구구.py
import sys
sys.setrecursionlimit(25000000)

n = int(input())

# 인접 리스트
adj = [[] for _ in range(n + 1)]

# 간선 정보 입력 (a - b : 가중치 c)
for _ in range(n - 1):
    a, b, c = map(int, input().split())
    adj[a].append((b, c))
    adj[b].append((a, c))

visited = [0] * (n + 1)
visited[1] = 1  # 시작 노드 방문으로 표시
max_dist = 0  # 트리에서 가장 먼 거리 저장

def dfs(m, dist):
    global max_dist
    global visited
    for i, j in adj[m]:  # i: 다음 노드, j: 간선 가중치
        if not visited[i]:
            max_dist = max(max_dist, dist + j)  # 현재까지 거리 최댓값 갱신
            visited[i] = 1
            dfs(i, dist + j)  # DFS 계속 탐색

# 루트(1번)에서 출발
dfs(1, 0)
print(max_dist)